src/rocksdb.rs

```diff
@@ -1,56 +1,238 @@
+        for _ in 0..NUM_RETRY {
+            e = Some(retry());
+            match e.as_ref().unwrap() {
+                Ok(_) => break,
+                Err(e) => log::warn!("Error during transaction execution {}", e)
+
+            }
+        }
+        e.unwrap().expect(&format!("Failed to retry operation for {} times", NUM_RETRY))
 
-impl<D: Borrow<DB>> DatabaseMut for RocksHandle<D> {
-    fn set(&mut self, key: H256, value: Option<&[u8]>) {
-        self.clear_cache();
-        let db = self.db.0.borrow();
-        if let Some(value) = value {
-            db.put(key, value)
-                .expect("Unable to put value into database");
-        } else {
-            db.delete(key).expect("Unable to delete key from database");
+    };
+}
+pub type RocksHandle<'a, D> = CachedHandle<RocksDatabaseHandle<'a, D>>;
+
+impl<'a, D: Borrow<DB>> DbCounter for RocksHandle<'a, D> {
+    // Insert value into db.
+    // Check if value exist before, if not exist, increment child counter.
+    fn gc_insert_node<F>(&self, key: H256, value: &[u8], mut child_extractor: F)
+    where
+        F: FnMut(&[u8]) -> Vec<H256>,
+    {
+        retry! {
+            let db = self.db.db.borrow();
+            let mut tx = db.transaction();
+            // let mut write_batch = WriteBatch::default();
+            if tx.get_for_update(key.as_ref(), EXCLUSIVE).map_err(|e|anyhow::format_err!("Cannot get key {}", e))?.is_none() {
+                let rlp = Rlp::new(&value);
+                let node = MerkleNode::decode(&rlp)?;
+                trace!("inserting node {}=>{:?}", key, node);
+                for hash in ReachableHashes::collect(&node, &mut child_extractor).childs() {
+                    self.db.increase(&mut tx, hash)?;
+                }
+
+                tx.put(key.as_ref(), value)?;
+                tx.commit()?;
+            }
+        }
+    }
+    fn gc_count(&self, key: H256) -> usize {
+        let db = self.db.db.borrow();
+        let mut tx = db.transaction();
+        self.db
+            .get_counter_in_tx(&mut tx, key)
+            .expect("Cannot read value") as usize
+    }
+
+    // atomic operation:
+    // 1. check if key counter didn't increment in other thread.
+    // 2. remove key if counter == 0.
+    // 3. find all childs
+    // 4. decrease child counters
+    // 5. return list of childs with counter == 0
+    fn gc_try_cleanup_node<F>(&self, key: H256, mut child_extractor: F) -> Vec<H256>
+    where
+        F: FnMut(&[u8]) -> Vec<H256>,
+    {
+        let db = self.db.db.borrow();
+        if self.db.counter_cf.is_none() {
+            return vec![];
+        };
+        retry! {
+            let mut nodes = vec![];
+            //TODO: retry
+            if let Some(value) = db.get(key.as_ref())?{
+                let mut tx = db.transaction();
+                let count = self.db.get_counter_in_tx(&mut tx, key)?;
+                if count > 0 {
+                    return Ok(vec![])
```

Comment:
Maybe `retry` macro can works with some `FnMut` instance? Manually typed signature helps to read this block.